Add `*.copy` support to `dope.sh`

cprecioso opened the pull request:
- Part of https://github.com/wasp-lang/open-saas/issues/516
- Part of #526

First step for #526, this PR adds support for binary files in the `app_diff`.

- In `diff.sh`, before generating a patchfile, we first check if the file is binary. If so, we just copy it to the `app_diff` with a `.copy` suffix
- In `patch.sh`, after applying all the `.diff` files, we then apply all the `.copy` files by just copying them to `app`, without the suffix

This functionality is not used in this PR, but in the next one #526

> [!NOTE]
> Made by 🤖 Claude and checked by 👨🏼 me

FranjoMindek reviewed (commented):
I'm gonna be a weak reviewer on here as Martin loves these.

The logic seems fine.
We only create diffs on difference i.e. no unnecessary duplication.

Nice initiative for making the template cleaner.

cprecioso commented on opensaas-sh/tools/dope.sh:
```diff
@@ -62,12 +62,31 @@ recreate_diff_dir() {
-      echo "${DIFF_OUTPUT}" > "${DIFF_DIR}/${filepath}.diff"
-      echo "Generated ${DIFF_DIR}/${filepath}.diff"
+    # Check if the file is binary
+    if file --mime "${derivedFilepath}" | grep -q "charset=binary"; then
```
It seems this `file --mime` incantation is the generally recommended way to check for binary-ness online.

What qualifies as binary? -> It seems to be a bit loose among each tool, but in this usecase it's mostly any file that consists of (or might) byte strings that are not well-formed text codepoints.

The NUL byte in particular seems to cause warnings and errors (both for bash pipes, and for the diff/patch utilities). The font files contain that byte, that's the main motivation for this change.

Martinsos replied:
Ok, makes sense! raw

cprecioso commented:
@Martinsos check again pls.
Refactored it a bit, extracted to named functions so that I could remove most of the comments (left two, as kind of headers)

Martinsos reviewed (approved):
Looking good! I approved, but please check some of the comments, there are some minor improvements -> once you are happy with those, feel free to merge, I don't have to look agian.

cprecioso commented on opensaas-sh/tools/dope.sh:
```diff
@@ -39,6 +39,25 @@ ACTION=$3
 DIFF_DIR="${DERIVED_DIR}_diff"
 DIFF_DIR_DELETIONS="${DIFF_DIR}/deletions"
 
+file_is_binary() {
+  local file=$1
+  if file --mime "$file" | grep -q "charset=binary"; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+files_are_equal() {
+  local file1=$1
+  local file2=$2
+  if cmp -s "$file1" "$file2"; then
+    return 0
+  else
+    return 1
+  fi
+}
+
```
> Why not just

AFAIU we need to return the input code, not the output of the function. I didn't want to do the somewhat-weird `echo $?` incantation, preferred being specific.

I'll test out both things in case I was mistaken and it works nonetheless.

cprecioso replied:
Bash does return the status code of the last run command inside a function, so the `if/return/else/return` was indeed excessive

Martinsos replied:
> > you can put these at the bottom of the file, out of the way, since they are helper utility functions.
> 
> I don't think so, no? Shell needs the functions defined before usage, right? https://unix.stackexchange.com/questions/449498/call-function-declared-below

Well, I remember that I defined utility functions at the bottom of installer.sh and then used them above. Ok oh, you are right, I just checked, so what I had there was this `main` function at top and then that was called at the very bottom, ok makes sense :). It is interpreted language with no second interpreter pass, so it can't know about func definitions if they aren not interpreted before the calls.
